### PR TITLE
feat(client): [internal] container extensions

### PR DESCRIPTION
**Pull request by jason-ha**
mimimal changes within existing packages to support distributed independent state experimental feature and other features that are container associated but not data stores.

Uses the proposed routing changes found in PR #22155.
Extension use address prefix `/ext/` and use will throw on attempt to send without "2.2" compatibility enabled.

JsonSerializable support is PR #21725 but used here.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -4493,3 +4523,57 @@ export class ContainerRuntime
 		return this.sessionSchema.opGroupingEnabled === true;
 	}
 }
+
+/**
+ * Represents the runtime of the container. Contains helper functions/state of the container.
+ * It will define the store level mappings.
+ */
+class ContainerRuntimeInternal extends ContainerRuntime implements IRuntimeInternal {
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/containerRuntime.ts:4531]: Please avoid class inheritance as the mechanism to add routes - at minimum this is challenging to scale to further extended functionality along with the usual downsides of class inheritance.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -2876,17 +2892,22 @@ export class ContainerRuntime
...
+
 		// Due to a mismatch between different layers in terms of
 		// what is the interface of passing signals, we need to adjust
 		// the signal envelope before sending it to the datastores to be processed
 		const envelope2: IEnvelope = {
-			address: envelope.address,
+			address,
 			contents: transformed.content,
 		};
 		transformed.content = envelope2;
```
Comment (ChumpChief) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2913]: Just below, we have this.channelCollection.processSignal().  Which is really just "read the envelope's address and forward to the correct recipient".

Maybe this would be a reasonable place to introduce your new routing, if it were generalized to something effectively like this.addressableRecipients.processSignal()?  I'm mostly thinking it would be nice if there's a single place that holds knowledge of valid addresses and the responsibility of routing to them.

**Review by ChumpChief — COMMENTED** [review]
Commenting just on the containerRuntime integration - I think a challenge of having a special "dis" address (whether through class extension or any other mechanism) is that whoever recognizes that address is effectively coupled to DIS.  If there's some approach to let DIS have an address, but not have it be "special" (instead just similar to any normal channel) it'd be easier to avoid the coupling.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (jason-ha) [reply packages/runtime/container-runtime/src/containerRuntime.ts:4531]: Open to alternative. ContainerRuntime is `@legacy` and none of the rest of this should be exposed (my current opinion).

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (jason-ha) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I've made an incremental step in that direction.
DataStore addressing and typing is fixed (not changable) and will remain exceptional.
And unless there is a resolution to what can be internal or not inheritance means logic will be separated.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:4536]: The only way for container runtime author to create a container is by using a concrete implementation of ContainerRuntime today. We can change that and expose a factory, and such factory could take a bunch of args to decide what functionality to mix in, but this does not exist today. Plus, I doubt that IContainerRuntime has all the right things exposed on it to be a viable replacement for ContainerRuntime type.

You can easily see the size of the problem by looking at Bohemia repo.

In other places we solved that same problem via mixins (i.e. dynamically created class types), allow arbitrary inheritance. It solves some problems, leaves other problems not solved. I personally prefer mixin approach overexposing 100 hooks, but I also see the opposite point of view (harder to see contracts, etc.).

File: packages/common/container-definitions/src/containerExtension.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	 * Client ID of the singular client the message is being (or has been) sent to.
+	 * May only be specified when IConnect.supportedFeatures['submit_signals_v2'] is true, will throw otherwise.
+	 */
+	targetClientId?: ClientConnectionId;
+}
+
+/**
+ * @internal
+ */
+export interface IContainerExtension<TContext extends unknown[]> {
```
Comment (vladsud) [inline packages/common/container-definitions/src/containerExtension.ts:51]: I did not spend that much time building mental model, but it's relatively hard to follow the responsibilities of each of these objects, and how they are used in a workflow. Having at least some comments above each interface / class providing some info like that would be super useful

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (jason-ha) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I admit to not understanding what is really supported with `ContainerRuntime` today and what we want.
As far as I can tell no one should create `ContainerRuntime` directly. They should be passing factory (`containerRuntimeCtor`) to `ContainerRuntime.loadRuntime`, no? I see 0 [hits for `containerRuntimeCtor` in Bohemia](https://office.visualstudio.com/OC/_search?action=contents&text=containerRuntimeCtor&type=code). Mixins are okay, but I still don't know what we want to expose or not. I really don't have a picture of where we are going; so, I can't fully judge if I am recommending an appropriate route.
While new feature I need support for remains experimental, I think it is okay to not have the full internal structure settled. It should be okay for someone doing something custom with `ContainerRuntime` to not get this support because is not settled.
All that said I switched to a mixin pattern here.

File: packages/common/container-definitions/src/containerExtension.ts (hunk 3 above)
Comment (jason-ha) [reply packages/common/container-definitions/src/containerExtension.ts:54]: I didn't spend time documenting the container extension part too much because I think it is everything surrounding it that matters more. Details matter so I have added more comments and hopefully the relationships are more clear now. The implementation of store in packages/runtime/container-runtime/src/containerRuntime.ts (mostly `acquireExtension` at line 4650) and the use of the feature in packages/runtime/container-runtime/src/test/containerExtensionStore.extension.example.ts's `hookItUp` are perhaps more helpful.
